### ISSUE TITLE
starving timeout with `async_io` on closed connections

### ISSUE THREAD

**Issue by flumm** [tokio-rs/tokio#6971]
**Version**

tokio v1.41.1
tokio-macros v2.4.0

**Platform**
Linux hostname 6.11.0-1-pve #1 SMP PREEMPT_DYNAMIC PMX 6.11.0-1 (2024-10-23T15:32Z) x86_64 GNU/Linux

**Description**
Is it expected behaviour that i can easily prevent a timeout from happening with 'async_io'  on closed connections?

I tried this code:

```rust
use std::{io, time::Duration};

use anyhow::Error;
use tokio::io::Interest;

#[tokio::main]
async fn main() -> Result<(), Error> {
    let server = tokio::net::TcpListener::bind("127.0.0.1:65000").await?;

    let (incoming, _socket) = server.accept().await?;
    eprintln!("connection established");

    let future = incoming.async_io(Interest::READABLE, || {
        Err::<(), io::Error>(io::ErrorKind::WouldBlock.into())
    });

    let _ = tokio::time::timeout(Duration::from_secs(5), future).await;

    eprintln!("timeout reached");

    Ok(())
}
```

(In our real code we call peek on the underlying socket to check if there is enough data available, and if not, we return WouldBlock)

After that, i connect to the server with `nc localhost 65000`

If i cancel the nc connection (with CTRL+C) before the 5 seconds, the timeout is never reached and it loops forever in the async_io callback
If i  don't close the connection (regardless if i send data or not), the timeout is reached normally.

I would have expected the timeout to trigger in both cases, but it does not.

**Comment by Darksonn** [comment]
Please share the actual code you are using. Your example `async_io` violates the API contract of `async_io` by returning `WouldBlock` without having received that error from the OS in a syscall on the socket.

**Comment by flumm** [comment]
the original code in the async_io callback is this:

```rust
let mut buf = [0; HANDSHAKE_BYTES_LEN];

let raw_fd = incoming_stream.as_raw_fd();
let std_stream =
unsafe { ManuallyDrop::new(std::net::TcpStream::from_raw_fd(raw_fd)) };

let peek_res = std_stream.peek(&mut buf);

match peek_res {
    Ok(peek_len) if peek_len < HANDSHAKE_BYTES_LEN => {
        Err(io::ErrorKind::WouldBlock.into())
    }
    res => res.map(|_| contains_tls_handshake_fragment(&buf)),
}
```

(see the original code: https://git.proxmox.com/?p=proxmox.git;a=blob;f=proxmox-rest-server/src/connection.rs;h=3815a8f40c082410c8e0091f4a57541f616003d2;hb=refs/heads/master#l473 )

the reason we do this is because we want to look at the data sent from the client and decide if it's a tls handshake, so we either continue with an encrypted connection or handle an unencrypted HTTP request

while i understand that this violates the contract of `async_io` it's still unexpected that this messes with the timeout in a way that this never finished, since the contract only talks about the socket in this case, not about how the future is polled/executed...
(it says:

> [...] which can cause the socket to behave incorrectly.

)

**Comment by Darksonn** [comment]
Yeah, that definitely violates the API contract. Unfortunately the OS provides no way to do what you want, so you're out of luck. It's not a Tokio limitation.

As for the issue itself, I guess the request is to fail more gracefully when the API contract is violated?

**Comment by Darksonn** [comment]
The problem is that `async_io` ends up [blocking the thread](https://ryhl.io/blog/async-what-is-blocking/).

**Comment by flumm** [comment]
> The problem is that `async_io` ends up [blocking the thread](https://ryhl.io/blog/async-what-is-blocking/).

ok while i get that, two things are still surprising here:

1. async_io returns a future that is seemingly easy to make blocking code, IMHO a future generated by a library should not indefinitely block, even when "holding wrong", or at least should mention so in the docs
2. having a timeout future that is not scheduled even when there are multiple idle threads available
 (i also tried a "manual" timeout future with sleep + select!, but that didn't work too)
 
 the questions are: should async_io "yield" (or the rust equivalent) between async_io callback calls ?
 and should tokio schedule the timeout future more intelligently?

**Comment by Darksonn** [comment]
The timeout future can't be scheduled for execution because it's already busy running. It's stuck inside [this call to `poll`](https://github.com/tokio-rs/tokio/blob/772e0ca8a678138ad4efa4d4b1f0ea544bca9bbe/tokio/src/time/timeout.rs#L202), which never returns as the future is blocking the thread. 

It is reasonable to make `async_io` behave less badly when you hold it wrong, though we shouldn't make correct usage worse by yielding unnecessarily.

**Comment by xXJDXx181** [comment]
Use this.    use std::{io, time::Duration};
use anyhow::Error;
use tokio::io::{AsyncReadExt, Interest};
use tokio::net::TcpListener;

#[tokio::main]
async fn main() -> Result<(), Error> {
    let server = TcpListener::bind("127.0.0.1:65000").await?;

    let (mut socket, _) = server.accept().await?;
    eprintln!("connection established");

    // Create a future that handles socket I/O, with timeout
    let future = tokio::select! {
        _ = tokio::time::timeout(Duration::from_secs(5), async {
            let mut buf = [0u8; 1024];
            loop {
                match socket.read(&mut buf).await {
                    Ok(0) => {
                        // Connection closed by the peer
                        eprintln!("connection closed");
                        return Ok(());
                    }
                    Ok(_) => {
                        // Data read successfully
                        eprintln!("data read");
                        return Ok(());
                    }
                    Err(e) => {
                        if e.kind() == io::ErrorKind::WouldBlock {
                            // Socket is not ready for reading, continue to next iteration
                            continue;
                        } else {
                            // Some other error occurred (e.g., connection reset)
                            return Err(e.into());
                        }
                    }
                }
            }
        }) => {
            // Timeout occurred
            eprintln!("timeout reached");
            Err(io
